### PR TITLE
Migrate almost everything to Bazel

### LINKED ISSUE #239: go fmt silently errors
/kind bug

**What steps did you take and what happened:**

Changing from `/hack/verify-gofmt.sh` to `go fmt ./pkg/...` now just swallows errors.

``` shell
➜ make fmt
go fmt ./pkg/... ./cmd/...
pkg/cloud/aws/services/certificates/certificates.go
➜
```

**What did you expect to happen:**

Should have exited with an error code

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

Go 1.10 and 1.11 have different opinions on whitespace alignment, which is causing
developer issues.

**Environment:**

- Cluster-api-provider-aws version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by randomvariable**
**What this PR does / why we need it**:
Adds linters. Using golangci lint as metalinter, although is used in other sub-projects does not fit well with Bazel as it expects to execute other linters in the GOPATH, whereas the former is statically compiled altogether.

This gets set up via a Bazel run command that will use the Go toolchain
that Bazel supplies rather than whatever is local to the computer.

Mock generation is now provided a custom rule that relies on `go_genrule` from repo-infra.
There's a hack here to copy `textflag.h` into the correct place in Bazel's copy of the Golang SDK. This is a recommended solution from https://github.com/bazelbuild/rules_go/issues/399 

All image and binary building is switched to Bazel. To maintain current semantics,
Make will then copy the artifacts into GOPATH/bin.

`go generate` is still used to perform the deepcopy as I found it to difficult to translate
into a build rule. It's pretty quick (< 6s) though, so is placed as a Make dependency.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #239, #292

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
Moves to Bazel is the primary build system
```

### CODE REVIEW

**Comment by randomvariable** [comment]
I need to recheck these bazel rules.

Also, wrong version of Go in CI for the fmt rules we prefer, so maybe time to bite the bullet and kill the non-Bazel tests.

**Comment by randomvariable** [comment]
Still one more thing to do.

**Comment by randomvariable** [comment]
This is ready for review. /assign @liztio

**Comment by liztio** [comment]
@randomvariable this looks super great! Thank you so much for doing this!

I don't quite understand those removed vendor files. Is that because they were moved into WORKSPACE?

**Comment by randomvariable** [comment]
It's mostly cosmetic. In the current state running gazelle displays this:
```
src/sigs.k8s.io/cluster-api-provider-aws ➜
bazel run gazelle
Starting local Bazel server and connecting to it...
INFO: Analysed target //:gazelle (39 packages loaded).
INFO: Found 1 target...
Target //:gazelle up-to-date:
bazel-bin/gazelle.bash
INFO: Elapsed time: 4.523s, Critical Path: 0.20s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
gazelle: unrecognized import path "a"
gazelle: unrecognized import path "a"
gazelle: unrecognized import path "domain.name/importdecl"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "nosuchpkg"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "nosuchpkg"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "referrers"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "lib"
gazelle: unrecognized import path "a"
gazelle: unrecognized import path "b"
gazelle: unrecognized import path "a"
```

By excluding some of the vendored mocks that message goes away as well as their BUILD files.

**Comment by chuckha** [comment]
@liztio https://github.com/kubernetes/test-infra/pull/9920 this replaces the makefile builds

**Comment by liztio** [comment]
@randomvariable does running `dep ensure` add them back in?

**Comment by randomvariable** [comment]
dep leaves the go files in, but Gazelle won't attempt to generate BUILD files for them.

Here's how `make dep-ensure` looks now (also downloading dep for the first time in this instance):
``` shell
src/sigs.k8s.io/cluster-api-provider-aws ➜
make dep-ensure
Repository is set up correctly
Checking status of dependencies
Loading:
Loading: 0 packages loaded
Loading: 0 packages loaded
Loading: 0 packages loaded
Loading: 0 packages loaded
currently loading:
Analyzing: target //:dep (1 packages loaded)
INFO: Build options have changed, discarding analysis cache.
INFO: SHA256 (https://github.com/golang/dep/archive/22125cfaa6ddc71e145b1535d4b7ee9744fefff2.zip) = c27b041e2a7fa9ddc555fef4c8c5e95392d1de1d51f35809e1c66be7f7d92be4
Analyzing: target //:dep (11 packages loaded)
Analyzing: target //:dep (11 packages loaded)
INFO: SHA256 (https://github.com/golang/dep/archive/22125cfaa6ddc71e145b1535d4b7ee9744fefff2.zip) = c27b041e2a7fa9ddc555fef4c8c5e95392d1de1d51f35809e1c66be7f7d92be4
INFO: SHA256 (https://github.com/golang/dep/archive/22125cfaa6ddc71e145b1535d4b7ee9744fefff2.zip) = c27b041e2a7fa9ddc555fef4c8c5e95392d1de1d51f35809e1c66be7f7d92be4
Analyzing: target //:dep (11 packages loaded)
INFO: SHA256 (https://codeload.github.com/golang/tools/zip/3e7aa9e59977626dc60433e9aeadf1bb63d28295) = 2384fa91351a7414b643c5230422ce45f5aa2be8a82727609afd4e64e6973a30
Analyzing: target //:dep (55 packages loaded)
INFO: Repository rule 'com_github_golang_protobuf' returned: {"remote": "https://github.com/golang/protobuf", "commit": "aa810b61a9c79d51363740d207bb46cf8e620ed5", "shallow_since": "2018-08-14", "init_submodules": False, "verbose": False, "strip_prefix": "", "patches": [Label("@io_bazel_rules_go//third_party:com_github_golang_protobuf-gazelle.patch"), Label("@io_bazel_rules_go//third_party:com_github_golang_protobuf-extras.patch")], "patch_tool": "patch", "patch_args": ["-p1"], "patch_cmds": [], "name": "com_github_golang_protobuf"}
INFO: SHA256 (https://codeload.github.com/google/protobuf/zip/48cb18e5c419ddd23d9badcfe4e9df7bde1979b2) = b6b42f90c60b54732f764ae875623a9b05e6eede064173c36c6fea12dd376cdd
INFO: Analysed target //:dep (66 packages loaded).
INFO: Found 1 target...
[0 / 4] [-----] BazelWorkspaceStatusAction stable-status.txt
[244 / 258] [-----] GoCompile external/com_github_golang_dep/internal/feedback/linux_amd64_stripped/go_default_library%/github.com/golang/dep/internal/feedback.a
Target //:dep up-to-date:
  bazel-bin/dep_script
  bazel-bin/dep
INFO: Elapsed time: 50.480s, Critical Path: 7.01s
INFO: 72 processes: 72 linux-sandbox.
INFO: Build completed successfully, 77 total actions
INFO: Running command line: bazel-bin/dep status
INFO: Build completed successfully, 77 total actions
Finished verifying dependencies


src/sigs.k8s.io/cluster-api-provider-aws ➜
git status
On branch metalinter
Your branch is up to date with 'origin/metalinter'.

nothing to commit, working tree clean
```

And on a forced run of `dep ensure`, not just `dep status`:

``` shell
src/sigs.k8s.io/cluster-api-provider-aws ➜
make dep-install
bazel run dep ensure
INFO: Analysed target //:dep (1 packages loaded).
INFO: Found 1 target...
Target //:dep up-to-date:
  bazel-bin/dep_script
  bazel-bin/dep
INFO: Elapsed time: 2.638s, Critical Path: 0.12s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
bazel run //:gazelle
INFO: Analysed target //:gazelle (24 packages loaded).
INFO: Found 1 target...
Target //:gazelle up-to-date:
  bazel-bin/gazelle-runner.bash
  bazel-bin/gazelle
INFO: Elapsed time: 6.019s, Critical Path: 2.85s
INFO: 28 processes: 28 linux-sandbox.
INFO: Build completed successfully, 33 total actions
INFO: Build completed successfully, 33 total actions
src/sigs.k8s.io/cluster-api-provider-aws ➜
git status
On branch metalinter
Your branch is up to date with 'origin/metalinter'.

nothing to commit, working tree clean
```

File: Makefile [hunk 1]
```diff
@@ -12,108 +12,155 @@
...
+
+.PHONY: clusterctl
+clusterctl: check-install generate lint ## Build clusterctl binary.
+	bazel build //cmd/clusterctl $(BAZEL_ARGS)
+	cp -f bazel-bin/cmd/clusterctl/*/clusterctl $(GOPATH)/bin/clusterctl
+
+.PHONY: clusterawsadm
+clusterawsadm: check-install dep-ensure ## Build clusterawsadm binary.
+	bazel build //cmd/clusterawsadm $(BAZEL_ARGS)
+	cp -f bazel-bin/cmd/clusterawsadm/*/clusterawsadm $(GOPATH)/bin/clusterawsadm
```
[the remark this answers is not in the document]
Comment (liztio) [reply Makefile]: I usually just run things with `bazel run //cmd/clusterawsadm`, and don't install them at all.

**Comment by chuckha** [comment]
@randomvariable I think this needs to be force pushed, but i just noticed there are 7 commits, could you squash this down to 1?

**Pull request merged**